pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,954 @@
+	allBGPPolicies, err := c.filterBGPPolicies()
+	if err != nil {
+		return err
+	}
+
+	allBGPPolicyNames := sets.KeySet(allBGPPolicies)
+	preState := c.bgpPolicyState
+	var curState *bgpPolicyState
+	var effectiveBGPPolicyName string
+	var needUpdateBGPServer bool
+
+	if preState == nil {
+		if allBGPPolicyNames.Len() == 0 {
+			return nil
+		}
+		// If there is no effective BGPPolicy in the last sync, select a random available BGPPolicy as the effective one.
+		effectiveBGPPolicyName, _ = allBGPPolicyNames.PopAny()
+		curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+		needUpdateBGPServer = true
+	} else {
+		// If there is an effective BGPPolicy in the last sync, check if it is still included in the available BGPPolicies.
+		if !allBGPPolicyNames.Has(preState.bgpPolicy) {
+			// If there is no available BGPPolicy, stop the previous BGP server, clean up the state and return.
+			if allBGPPolicyNames.Len() == 0 {
+				if err := preState.bgpServer.Stop(c.ctx); err != nil {
+					return err
+				}
+				c.bgpPolicyState = nil
+				return nil
+			}
+			// Select a new effective BGPPolicy from the available ones.
+			effectiveBGPPolicyName, _ = allBGPPolicyNames.PopAny()
+			curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+			needUpdateBGPServer = true
+		} else {
+			// Retain the effective BGPPolicy in the last sync.
+			effectiveBGPPolicyName = preState.bgpPolicy
+			curState = &bgpPolicyState{bgpPolicy: effectiveBGPPolicyName}
+		}
+	}
+	bgpPolicy := allBGPPolicies[effectiveBGPPolicyName]
+
+	klog.V(2).InfoS("Syncing BGPPolicy", "BGPPolicy", effectiveBGPPolicyName)
+	// Retrieve the listen port, local AS number and router ID from the effective BGPPolicy, and update them to the
+	// current state.
+	curState.routerID, err = c.getRouterID()
+	if err != nil {
+		return err
+	}
+	if bgpPolicy.Spec.ListenPort != nil {
+		curState.listenPort = *bgpPolicy.Spec.ListenPort
+	} else {
+		curState.listenPort = defaultBGPListenPort
+	}
+	curState.localASN = bgpPolicy.Spec.LocalASN
+
+	// If any of the listen port, local AS number, or router ID have changed, mark the BGP server for an update.
+	if preState != nil {
+		needUpdateBGPServer = needUpdateBGPServer ||
+			preState.listenPort != curState.listenPort ||
+			preState.localASN != curState.localASN ||
+			preState.routerID != curState.routerID
+	}
+
+	if needUpdateBGPServer {
+		// Stop the stale BGP server.
+		if preState != nil {
+			if err := preState.bgpServer.Stop(c.ctx); err != nil {
+				klog.ErrorS(err, "failed to stop stale BGP server")
+			}
+		}
+		// Start the new BGP server.
+		globalConfig := &bgp.GlobalConfig{
+			ASN:        uint32(curState.localASN),
+			RouterID:   curState.routerID,
+			ListenPort: curState.listenPort,
+		}
+		bgpServer := c.newBGPServerFn(globalConfig)
+		if err := bgpServer.Start(c.ctx); err != nil {
+			return err
+		}
+		// Update the current effective BGPPolicy state.
+		curState.bgpServer = bgpServer
```

Comment:
This instance would never be stopped if any of the following code fails, causing various resource leak.

Reply:
Will consider how to avoid that.

Reply:
Done